File: src/Application/Client/Java/ByteKeyValueIterator.java
package com.scalien.scaliendb;

import java.util.Collections;
import java.util.LinkedList;
import java.util.Map;
import java.util.TreeMap;

public class ByteKeyValueIterator implements java.lang.Iterable<KeyValue<byte[], byte[]>>, java.util.Iterator<KeyValue<byte[], byte[]>>
{
    private Table table;
    private byte[] startKey;
    private byte[] endKey;
    private byte[] prefix;
    private int count;
    private boolean forwardDirection;
    private int gran = 100;
    private int pos;
    private LinkedList<byte[]> keys;
    private LinkedList<byte[]> values;
 
    public ByteKeyValueIterator(Table table, ByteRangeParams ps) throws SDBPException {
        this.table = table;
        this.startKey = ps.startKey;
        this.endKey = ps.endKey;
        this.prefix = ps.prefix;
        this.count = ps.count;
        this.forwardDirection = ps.forwardDirection;
        
        query(false);
    }

    // for Iterable
    public java.util.Iterator<KeyValue<byte[], byte[]>> iterator() {
        return this;
    }
    
    // for Iterator
    public boolean hasNext() {
        try {
            if (count == 0)
                return false;
            if (pos == keys.size()) {
                if (keys.size() < gran)
                    return false;
                startKey = keys.get(keys.size()-1);
                query(true);
            }
            if (keys.size() == 0)
                return false;
            return true;
        }
        catch(SDBPException e) {
            return false;
        }
    }
    
    // for Iterator
    public KeyValue<byte[], byte[]> next() {
        KeyValue<byte[], byte[]> e = new KeyValue<byte[], byte[]>(keys.get(pos), values.get(pos));
        pos++;
        if (count > 0)
            count--;
        return e;
    }
    
    // for Iterator
    public void remove() throws UnsupportedOperationException {
        throw new UnsupportedOperationException();
    }
    
    private void query(boolean skip) throws SDBPException {
        int num;
        Map<byte[], byte[]> result;

        num = gran;
        if (count > 0 && count < gran)
            num = count;

        result = table.getClient().listKeyValues(
         table.getTableID(), startKey, endKey, prefix, num, forwardDirection, skip);
        keys = new LinkedList<byte[]>();
        values = new LinkedList<byte[]>();

        for (Map.Entry<byte[], byte[]> entry : result.entrySet()) {
            byte[] key = entry.getKey();
            byte[] value = entry.getValue();
            keys.add(key);
            values.add(value);
        }

        pos = 0;
    }    
}

File: src/Application/Client/CSharp/ScalienClient/ByteKeyValueIterator.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace Scalien
{
    public class ByteKeyValueIterator : IEnumerable<KeyValuePair<byte[], byte[]>>, IEnumerator<KeyValuePair<byte[], byte[]>>
    {
        Table table;
        bool forwardDirection; 
        byte[] startKey;
        byte[] endKey;
        byte[] prefix;
        long count;
        uint granularity = 100;
        int pos;
        List<byte[]> keys;
        List<byte[]> values;

        public ByteKeyValueIterator(Table table, ByteRangeParams ps)
        {
            this.table = table;
            this.startKey = ps.startKey;
            this.endKey = ps.endKey;
            this.prefix = ps.prefix;
            this.count = ps.count;
            this.granularity = ps.granularity;
            this.forwardDirection = ps.forwardDirection;
            Query(false);
        }

        private void Query(bool skip)
        {
            uint num;
            Dictionary<byte[], byte[]> result;

            num = granularity;
            if (count > 0 && count < granularity)
                num = (uint)count;
            
            result = table.Client.ListKeyValues(table.TableID, startKey, endKey, prefix, num, forwardDirection, skip);

            keys = new List<byte[]>();
            values = new List<byte[]>();

            foreach (KeyValuePair<byte[], byte[]> kv in result)
            {
                keys.Add(kv.Key);
                values.Add(kv.Value);
            }

            pos = 0;
        }

        #region IEnumerable interface

        IEnumerator IEnumerable.GetEnumerator()
        {
            return (IEnumerator)GetEnumerator();
        }

        public virtual IEnumerator<KeyValuePair<byte[], byte[]>> GetEnumerator()
        {
            return this;
        }

        #endregion

        #region IEnumerator interface

        public virtual bool MoveNext()
        {
            if (count == 0)
                return false;
            if (pos == keys.Count)
            {
                if (keys.Count < granularity)
                    return false;
                startKey = keys[keys.Count - 1];
                Query(true);
            }
            if (keys.Count == 0)
                return false;

            pos++;
            if (count > 0)
                count--;
            return true;
        }

        public virtual void Reset()
        {
            throw new NotSupportedException();
        }

        public virtual KeyValuePair<byte[], byte[]> Current
        {
            get
            {
                return new KeyValuePair<byte[], byte[]>(keys[pos - 1], values[pos - 1]);
            }
        }

        object IEnumerator.Current
        {
            get
            {
                return Current;
            }
        }

        void IDisposable.Dispose()
        {
        }

        #endregion
    }
}
